src/crush/mapper.c

```diff
@@ -833,6 +928,25 @@ int crush_do_rule(const struct crush_map *map,
 	w = a;
 	o = b;
 
+	if (map->reweight_items) {
+		/* This resets the weight at rule start. I think we
+		   dont allow collissions anywhere in CRUSH
+		   output. If there's some break after which we DO
+		   allow collissions, then the reset should be there,
+		   too. */
+
+		/* This and other reweighting modifications are
+		   externally invisible, so this should be fine. I'm
+		   following the example of the 'perm' array here
+		   which is modified in the course of mapping. */
+		reset_reweight((struct crush_map*)map);
+		/* Instead of modifying the map in-place we could
+		   have builder.c allocate a bookkeeping map, which
+		   is what I would have done if Perm hadn't already
+		   suggested we can keep information like this in
+		   place. */
```

Comment:
FWIW the fact that perm does this means we have to protect crush mappings with a big mutex, which sort of sucks and should really be fixed anyway.  Unfortunately it means we need to build a mapping structure of some sort.  Since crush choices are generally pretty simple (not that many buckets) perhaps a simple flat vector with O(n) lookups is fine?  We're stuck with straight C and can't easily use the kernel rbtree or similar structures since we share the code with userspace.  :/

In any case, though, I think we should do this one the "right" way with a local map so that the crush mapping function can be eventually made const.